### ISSUE TITLE
Need the ability to add an extra project to the solution as part of features

### ISSUE THREAD

**Issue by mrlacey** [Microsoft/WindowsTemplateStudio#916]
Need the ability to add an additional project to the generated solution.
As part of a feature, the template would need the ability to be able to add to and modify (postaction-merge) both the original project and the new one.

Implications for Right-Click>Add need to be explored and understood for this before implementing. Features that want to add extra projects shouldn't be prevented from being included as right click options.

Working on the assumption that the added project would be a library or set of tests and so couldn't later be extended by having extra things added to them through rightclick-add.
Any features that add another project to the solution would exist in isolation and so we wouldn't support features adding things to secondary projects. Unless there's a good reason why this shouldn't (at least initially) be the case.

This is needed by features which will add testing capabilities to the generated code (X-Ref #299 & #300) and possibly more.

Xamarin PoC has the ability to produce a solution with multiple projects in the generated solution but not explored if this can be done optionally. I presume this would need

**Comment by mrlacey** [comment]
For reference.
There is a way to do this as a postaction in the config file but this is for the generation of new projects, not the addition of one based on a defined template.

Docs: https://github.com/dotnet/templating/wiki/Post-Action-Registry#add-a-reference-to-a-project-file
Example: https://github.com/dotnet/templating/issues/601

I have looked at the Xamarin PoC branch but that just shows generating a multi-project solution. It doesn't allow adding a project to a solution through composition.

**Comment by mrlacey** [comment]
This needs to: 
- add the new project to the solution
- Have the new project add a reference to the existing project (and so will need the name of the existing project)
- Update code files in the new project based on the name of the existing project.

All this at least in the initial version.

In future, we'll also need to be able to:
- Have existing project reference the newly added one
- Make changes to the existing project based on the name of the new one (and other classes, etc. in it)
- Allow for specifying the name of the added project (possibly)

Shouldn't allow for circular referencing of added project. Reference should be one way or the other.

**Comment by mrlacey** [comment]
Removing myself as assignee as don't want to block others from moving this forward.

**Comment by mrlacey** [comment]
Moving to backlog as this is only needed to unblock #299 & #300 and both of them are low priority (for now)

**Comment by crutkas** [comment]
This should be there with xamarin as well

**Comment by mrlacey** [comment]
I hadn't thought about how this impacts Xamarin as I was considering that would just be adding multiple projects as part of the generation in the form of multiple heads. However, if heads are optional and right-click allows adding new heads then this will be necessary. In fact, allowing right-click to add new heads could be super complicated as it may need to know about all pages and features added previously.

There are no issues about Xamarin support but there is an experimental(?) [branch](https://github.com/Microsoft/WindowsTemplateStudio/tree/xamarin) to look at it but hasn't been touched in 3+ months.
If someone is working on Xamarin support and/or has spent time thinking about how it will work with right-click it'd be good to sync up and see how those thoughts relate to what I'm thinking about this issue and if it's going to require additional, related functionality.

**Comment by ralarcon** [comment]
We could think in adding other projects as a "Libraries" where the user can put pages / features but this will be a bit difficult as more than exactly knowing the already added, it will be required to make post-actions affecting the main (entry point) project (i.e, to register a page in the navigation service). Not sure if this scenario is really worthy.

Regarding Xamaring, our initial thoughts were to have the multiple projects generated in the "new project", the right-click actions targeting the shared project and minimize the modifications over the platform specific projects. But this is something we need to further investigate.

**Comment by mrlacey** [comment]
I've been looking at this as I want a way to run tests on equivalent versions of projects generated in C# & VB to confirm they behave the same way.
The Templating CLI includes a way via [post actions](https://github.com/dotnet/templating/wiki/Post-Action-Registry) to "Add a reference to a project file" and "Add projects to a solution file". 
Unfortunately, this functionality is only in the CLI and not in the core templating libraries that we use.

My plan, therefore, is to recreate simplified versions of these post actions, based on the CLI code ([here](https://github.com/dotnet/templating/blob/59ea5050650f0928e557d5fdaa80e894448ec671/src/Microsoft.TemplateEngine.Cli/PostActionProcessors/AddReferencePostActionProcessor.cs) and [here](https://github.com/Microsoft/WindowsTemplateStudio/blob/Issue863-VBTemplates/code/src/UI/ViewModels/NewProject/ProjectSetupViewModel.cs)) so we can have WTS functionality that is similar to the templating CLI.
This seems the best way to get something that is close to the command line "[dotnet-add reference](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-add-reference)" behavior.

I'm not yet sure how close what we need will match the actual CLI code but I'd like it to be semantically similar even if it's not identical. There may also be some simplifications that we can use too, such as using a meaningful string for an Id, rather than a GUID. We can also probably do without ContinueOnError & Description properties as we always break on an error and the description won't be seen anywhere. It'll also only work with referencing other projects as we have other methods for referencing packages and frameworks.

The projects that are to be added will be in subfolders of the "templating" folder on disk but won't have their own templating.config.

@sibille As you seem to have a good understanding of the core templating code, how does this sound? Have I missed something more obvious? Or, is there a better way?

**Comment by sibille** [comment]
@mrlacey When we started using the template engine their postactions were still under definition. So we decided to implement our own postactions and later evaluate if we can merge with their system. 

You can find all postactions available in WTS [here](https://github.com/Microsoft/WindowsTemplateStudio/tree/dev/code/src/Core/PostActions/Catalog). There is already a postaction to add a project to a solution, it's executed for each template with wts.type=project, and it adds the PrimaryOutput to the solution file. (There is a limitation from the VSEmulator project, as the FakeSolution does not add multiple projects correctly, but from the Installer.sln it's working fine, and VSEmulator will get a fix soon ;) 
So I think this one should work for you, although you'll have to add the project to a template.json with wts.type = project, is that a possibility?

For the "Add a reference to a project file" there's no implementation from our side currently. 

We're currently having a look at the template engine postaction definition to evaluate if we can use their postaction system. Xamarin projects require us to have postactions that can be defined by the template instead of the template type. (For example the GenerateTestCertificatePostAction should only be executed for the UWP project, not for the other ones).

Let's sync when we have a bit more insight and see how we can accomplish this.

**Comment by ralarcon** [comment]
@mrlacey We have been reviewing the post-action in the template engine and, as you already mentioned, the postaction implementation depends on the host executing the generation.

What I saw is that we can more or less easily hook up in the Post-Action infrastructure so we can have post-actions allowing to be specified by the templates.

For example, for the xamarin stuff we will need that only the Uwp project to add the pfx certificate. This is a post-action that currently works at project level (everytime a project is generated) so I think this make sense to be defined in the template instead of executing it for each project. The "Add a reference to a project file" make sense to be defined in the template too.  In the other hand, the global post-actions, the merge post-actions and some others makes sense to do NOT specify at template level. 

In summary, we can have a mix of post-actions types. I'm going to try to change the post-action which generate the certificate and discover exaclty what is needed.

**Comment by mrlacey** [comment]
removing assignment as others are working in this area and it might change dramatically.

**Comment by mrlacey** [comment]
Adding to 2.1 as most of this is already in anyway.

**Comment by mrlacey** [comment]
Additional changes merged today (on the same PR as previous) mean this should be good now.

Implementing #300 will prove this all works as expected.

**Comment by mrlacey** [comment]
This all works fine when running through Visual Studio.
`FakeSolution` (as used by VSEmulator) will need extending to support different project types. As the exact changes required are likely to vary by project type, I propose to make those modifications as and when they are needed for the features, etc. we add.
I'm, therefore, going to close this. Feel free to reopen and we can discuss this further if you wish.
